### ISSUE TITLE
Support OPA integrationq

### ISSUE THREAD

**Issue by shohamyamin** [hansetag/iceberg-catalog#266]
Is it possible to support opa integration to the catalog.
I found that implementing ABAC with openFGA is not that easy

**Comment by c-thiel** [comment]
Hi @shohamyamin, in principle it is possible and we do want to be open for other forms of Auth. However openFGA has a higher priority for us and we will probably change the current interface for AuthZ in the next month.

Do you have a few more details on how you envision ABAC? Would attributes be stored as part of iceberg-metadata, or would they be stored separately? I am asking mainly because openFGA is pretty much a self-contained dynamic system, while working with OPA is typically dependent on an external dynamic store (i.e. Attributes of Tables as part of Iceberg Metadata) and OPA is a quite static rule processing engine - you wouldn't store Attribute <- Table Mappings in OPA directly.

**Comment by shohamyamin** [comment]
Yes. The case of abac that i am taking about is when you have attributes for both the user and the resources(tables/columns) so if it is possible that this will come from the catalog itself( the user attribute from the connection to the AD and the table asttributes some how to save it in the catalog) as part of the input of the request to opa that whould be great but even if opa will be needed to take care of getting the attributes from diffrent source that still will be good. in opa there are a lot of diffrent ways to load data to the opa. there is thing that called opal that has api/postgres/mongo fetcher that allows the opa to fetch the data every X minutes?

there is also the possibilty to fetch the data from with simple http request as part of the opa policy  

The opa integration for start should be simple only providing the information about the action that is being performed and the resources that relevents for that action and the all logics will be in the opa policies.


Here is how Trino implements an OPA plugin for some reference:
https://trino.io/docs/current/security/opa-access-control.html#security-opa-access-control--page-root

**Comment by c-thiel** [comment]
@shohamyamin we now have a first integration with OPA.

Currently the direction is:
1. trino sends a request to OPA
2. OPA translate the request to a Lakekeeper permission check (its almost the same API as trino OPA)
3. OPA sends the request to Lakekeeper, which checks the permissions. Permission tuples live in Postgres via OpenFGA
4. OPA allows / disallows the operation depending on the response. 

Docs are here: https://docs.lakekeeper.io/docs/nightly/opa/ and we also have a standalone docker compose example: https://github.com/lakekeeper/lakekeeper/tree/main/examples/trino-opa

What we don't have yet is the other way around - where Lakekeeper asks OPA for permissions.

**Comment by shohamyamin** [comment]
@c-thiel thats great news.

The reverse approach, where Lakekeeper queries OPA, would also be excellent as it offers greater flexibility in the authorization logic.

Another concern I had is whether the current implementation results in a request to Lakekeeper for every request made to OPA. This could potentially impact query performance, as each query to Trino could involve multiple OPA requests

**Comment by c-thiel** [comment]
@shohamyamin yes, every OPA request is forwarded. This is required as security is a sensitive topic and we want point in time revocation of access. The impact on performance, while not beeing 0, is very low, because:
- Lakekeeper uses OpenFGA under the hood, which is smart about caching and answers requests in single digit ms
- Lakekeepers OPA bridge supports trino batch requests - thus many requests from trino are processed in parallel
- On Kubernetes, we can easily co-locate OPA and Lakekeeper to effectively eliminate network latency

I see your point though with an external OPA. Would you also be willing to contribute it? Or a part of it?
